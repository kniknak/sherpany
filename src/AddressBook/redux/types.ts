export enum Status {
    Initial = 1,
    Success,
    Pending,
    Failure,
}

export type Nationality = "CH" | "ES" | "FR" | "GB"

export interface AddressBookState {
    status: Status
    users: FetchAddressBookUser[][]
    page: number
    nationalities: Set<Nationality>
}

export interface FetchAddressBookUser {
    name: {
        first: string
        last: string
    }
    location: {
        street: {
            name: string
            number: number
        }
        city: string
        state: string
        postcode: string
    }
    email: string
    login: {
        username: string
    }
    phone: string
    cell: string
    id: {
        value: string
    }
    picture: {
        thumbnail: string
    }
}

export interface FetchAddressBookPayload {
    users: FetchAddressBookUser[]
    page: number
}
