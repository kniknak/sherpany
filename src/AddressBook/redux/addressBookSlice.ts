import { createSlice, PayloadAction } from "@reduxjs/toolkit";
import { AddressBookState, FetchAddressBookPayload, Nationality, Status } from "./types";

export const nationalities: Nationality[] = ["CH", "ES", "FR", "GB"];

export const addressBookInitialState: AddressBookState = {
    status: Status.Initial,
    users: [],
    page: 1,
    nationalities: new Set(nationalities),
};

export const addressBookSlice = createSlice({
    name: "counter",
    initialState: addressBookInitialState,
    reducers: {
        request: state => {
            state.status = Status.Pending;
        },
        success: (state, action: { payload: FetchAddressBookPayload }) => {
            state.status = Status.Success;
            state.users[action.payload.page - 1] = action.payload.users;
        },
        failure: state => {
            state.status = Status.Failure;
        },
        toggleNationality: {
            reducer: (state, action: PayloadAction<Nationality>) => {
                if (state.nationalities.has(action.payload)) {
                    state.nationalities.delete(action.payload);
                } else {
                    state.nationalities.add(action.payload);
                }
            },
            prepare: (nationality: Nationality) => ({ payload: nationality }),
        },
    },
});