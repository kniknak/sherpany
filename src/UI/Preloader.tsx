import React, {memo} from "react";
import Typography from "@material-ui/core/Typography";
import Toolbar from "@material-ui/core/Toolbar";

export const Preloader: React.FC = memo(({ children }) => {
    return (
        <Toolbar>
            <Typography
                align="center"
                style={{ flex: 1 }}
            >
                {children}
            </Typography>
        </Toolbar>
    );
});