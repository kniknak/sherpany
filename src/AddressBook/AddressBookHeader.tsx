import React from "react";
import { RootState } from "../App/store";
import Typography from "@material-ui/core/Typography";
import Toolbar from "@material-ui/core/Toolbar";
import AppBar from "@material-ui/core/AppBar";
import MuiTextField from "@material-ui/core/TextField";
import withStyles from "@material-ui/core/styles/withStyles";
import { Status } from "./redux/types";
import { connect, MapStateToProps } from "react-redux";

const TextField = withStyles(theme => ({
    root: {
        width: 100,
        marginLeft: theme.spacing(2),
    },
}))(MuiTextField);

interface StateProps {
    status: Status
}

interface OwnProps {
    status: Status
    search: string
    setSearch: (search: string) => void
}

export type AddressBookHeaderProps = StateProps & OwnProps

export class AddressBookHeader extends React.PureComponent<AddressBookHeaderProps> {
    private onChange: React.ChangeEventHandler<HTMLInputElement> = event => this.props.setSearch(event.target.value)

    render() {
        const { status, search } = this.props

        return (
            <AppBar position="sticky"
                color="default">
                <Toolbar style={{ alignItems: "baseline" }}>
                    <Typography variant="body1">
                        {Status[status]}
                    </Typography>
                    <Typography
                        variant="body1"
                        style={{ flex: 1 }}
                        align="right"
                    >
                        Search:
                    </Typography>
                    <TextField
                        label="Type text"
                        value={search}
                        onChange={this.onChange}
                        margin="normal"
                        variant="outlined"
                        size="small"
                    />
                </Toolbar>
            </AppBar>
        );
    }
}


const mapStateToProps: MapStateToProps<StateProps, {}, Pick<RootState, "addressBook">> = state => ({
    status: state.addressBook.status,
});

export const AddressBookHeaderConnected = connect(mapStateToProps)(AddressBookHeader);

