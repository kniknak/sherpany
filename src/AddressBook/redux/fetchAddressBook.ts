import { createAction } from "redux-api-middleware";
import { addressBookSlice, nationalities } from "./addressBookSlice";
import { FetchAddressBookPayload, FetchAddressBookUser } from "./types";
import { RootState } from "../../App/store";

export const fetchAddressBook = (page: number) => createAction<Pick<RootState, "addressBook">, Promise<FetchAddressBookPayload>, never>({
    endpoint: state => {
        const nat = [...(state.addressBook.nationalities.size
            ? state.addressBook.nationalities
            : nationalities),
        ]
            .join(",");

        return `https://randomuser.me/api/?seed=sheerpany&page=${page}&results=50&nat=${nat}`;
    },
    method: "GET",
    types: [
        addressBookSlice.actions.request.type,
        {
            type: addressBookSlice.actions.success.type,
            payload: (action, state, response) => (
                response
                    .json()
                    .then<FetchAddressBookPayload>((result: FetchAddressBookResponse) => ({
                        users: result.results,
                        page,
                    }))
            ),
        },
        addressBookSlice.actions.failure.type,
    ],
});

interface FetchAddressBookResponse {
    results: FetchAddressBookUser[]
    info: {
        seed: "sheerpany"
        results: number
        page: number
        version: string
    }
}