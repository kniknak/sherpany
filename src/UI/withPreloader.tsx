import React, { Suspense } from "react";
import { Preloader } from "./Preloader";

export const withPreloader = <TProps extends object = {}>(Component: React.ComponentType<TProps>): React.ComponentType<TProps> => (
    (props) => (
        <Suspense fallback={<Preloader>loading...</Preloader>}>
            <Component {...props}/>
        </Suspense>
    )
);