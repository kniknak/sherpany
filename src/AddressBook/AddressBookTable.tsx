import React from "react";
import { RootState } from "../App/store";
import { connect, MapStateToProps } from "react-redux";
import Table from "@material-ui/core/Table";
import TableHead from "@material-ui/core/TableHead";
import TableRow from "@material-ui/core/TableRow";
import TableCell from "@material-ui/core/TableCell";
import TableBody from "@material-ui/core/TableBody";
import { AddressBookPerson } from "./Person/AddressBookPerson";
import throttle from "lodash.throttle";
import { fetchAddressBook } from "./redux/fetchAddressBook";
import { FetchAddressBookUser, Status } from "./redux/types";
import { Preloader } from "../UI/Preloader";
import { AddressBookPersonModal } from "./Person/AddressBookPersonModal";

const isPersonShown = (person: FetchAddressBookUser, search: string) => (search === "") || (
    [person.name.first, person.name.last]
        .join(" ")
        .toLowerCase()
        .includes(search.toLowerCase())
);

interface StateProps {
    users: FetchAddressBookUser[][]
    status: Status
}

interface DispatchProps {
    fetchAddressBook: (page: number) => void
}

interface OwnProps {
    search: string
    scrollThreshold: number
    maxPage: number

}

export type AddressBookTableProps = StateProps & DispatchProps & OwnProps

interface State {
    page: number
    isOpen: boolean
    person?: FetchAddressBookUser
}

export class AddressBookTable extends React.PureComponent<AddressBookTableProps, State> {
    private throttleInterval = 500;

    readonly state: State = {
        page: 1,
        isOpen: false,
    };

    componentDidMount() {
        window.addEventListener("scroll", this.onScroll);
        this.fetchResults(this.state.page);
    }

    componentWillUnmount() {
        window.removeEventListener("scroll", this.onScroll);
    }

    private onScroll = () => {
        if (this.props.search !== "") {
            return;
        }

        if (window.innerHeight + document.documentElement.scrollTop >= document.documentElement.offsetHeight - this.props.scrollThreshold) {
            if (this.props.status !== Status.Pending) {
                const page = Math.min(this.props.maxPage, this.state.page + 1);
                this.setState({ page });
                this.fetchResults(page);
            }
        }

    };

    private fetchResults = throttle((page: number) => this.props.fetchAddressBook(page), this.throttleInterval);

    private onOpen = (person: FetchAddressBookUser) => this.setState({
        isOpen: true,
        person,
    });

    private onClose = () => this.setState({
        isOpen: false,
        person: undefined,
    });

    private renderPage = (slice: FetchAddressBookUser[], page: number) => (
        <React.Fragment key={page}>
            {slice.map(this.renderPerson)}
        </React.Fragment>
    );

    private renderPerson = (person: FetchAddressBookUser) => (
        <AddressBookPerson
            key={person.id.value}
            person={person}
            isShown={isPersonShown(person, this.props.search)}
            onOpen={this.onOpen}
        />
    );

    render() {
        const { users, maxPage, status } = this.props;
        const { page, isOpen, person } = this.state;

        return (
            <>
                <Table>
                    <TableHead>
                        <TableRow>
                            <TableCell>picture.thumbnail</TableCell>
                            <TableCell>name.first</TableCell>
                            <TableCell>name.last</TableCell>
                            <TableCell>name.username</TableCell>
                            <TableCell>email</TableCell>
                            <TableCell/>
                        </TableRow>
                    </TableHead>
                    <TableBody>
                        {users.map(this.renderPage)}
                    </TableBody>
                </Table>
                {status === Status.Pending ? (
                    <Preloader>loading...</Preloader>
                ) : (
                    (page === maxPage) && (
                        <Preloader>end of users catalog</Preloader>
                    )
                )}
                <AddressBookPersonModal
                    person={person}
                    isOpen={isOpen}
                    onClose={this.onClose}
                />
            </>
        );
    }
}

const mapStateToProps: MapStateToProps<StateProps, OwnProps, Pick<RootState, "addressBook">> = state => ({
    users: state.addressBook.users,
    status: state.addressBook.status,
});

const mapDispatchToProps: DispatchProps = {
    fetchAddressBook,
};
export const AddressBookTableConnected = connect(mapStateToProps, mapDispatchToProps)(AddressBookTable);
