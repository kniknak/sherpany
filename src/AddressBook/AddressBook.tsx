import React, { useState, memo } from "react";
import { AddressBookTableConnected } from "./AddressBookTable";
import { AddressBookHeaderConnected } from "./AddressBookHeader";

const scrollThreshold = 500;
const maxPage = 20;

export const AddressBook: React.FC = memo(() => {
    const [search, setSearch] = useState("");

    return (
        <>
            <AddressBookHeaderConnected
                search={search}
                setSearch={setSearch}
            />
            <AddressBookTableConnected
                search={search}
                scrollThreshold={scrollThreshold}
                maxPage={maxPage}
            />
        </>
    );
});

export default AddressBook;