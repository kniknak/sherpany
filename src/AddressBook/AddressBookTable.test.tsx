import React from "react";
import { shallow } from "enzyme";
import { AddressBook } from "./AddressBook";
import { AddressBookTable, AddressBookTableProps } from "./AddressBookTable";
import { FetchAddressBookUser, Status } from "./redux/types";
import { Preloader } from "../UI/Preloader";
import TableBody from "@material-ui/core/TableBody";

describe("AddressBook", () => {
    describe("AddressBookTable", () => {
        const person: FetchAddressBookUser = {
            name: {
                first: "name.first",
                last: "name.last",
            },
            location: {
                street: {
                    name: "location.street.name",
                    number: 1,
                },
                city: "location.city",
                state: "location.state",
                postcode: "location.postcode",
            },
            email: "email",
            login: {
                username: "login.username",
            },
            phone: "phone",
            cell: "cell",
            id: {
                value: "id.value",
            },
            picture: {
                thumbnail: "picture.thumbnail",
            },
        };
        const person2: FetchAddressBookUser = {
            ...person,
            id: {
                value: "id.value2",
            },
        };
        const person3: FetchAddressBookUser = {
            ...person,
            id: {
                value: "id.value2",
            },
        };
        const person4: FetchAddressBookUser = {
            ...person,
            id: {
                value: "id.value2",
            },
        };
        const defaultProps: AddressBookTableProps = {
            search: "",
            scrollThreshold: 0,
            maxPage: 1,
            fetchAddressBook: () => undefined,
            status: Status.Success,
            users: [],
        };

        it("renders", () => {
            const subject = shallow(<AddressBookTable {...defaultProps}/>);

            expect(subject).toMatchSnapshot();
        });

        it("renders loading state", () => {
            const props: AddressBookTableProps = {
                ...defaultProps,
                status: Status.Pending,
            };

            const subject = shallow(<AddressBookTable {...props}/>);

            expect(subject.find(Preloader)).toHaveLength(1);
            expect(subject.find(Preloader)).toMatchSnapshot();
        });

        it("renders end state", () => {
            const props: AddressBookTableProps = {
                ...defaultProps,
                status: Status.Success,
                maxPage: 1,
            };

            const subject = shallow(<AddressBookTable {...props}/>);

            subject.setState({ page: props.maxPage });

            expect(subject.find(Preloader)).toHaveLength(1);
            expect(subject.find(Preloader)).toMatchSnapshot();
        });

        it("renders table", () => {
            const props: AddressBookTableProps = {
                ...defaultProps,
                status: Status.Success,
                users: [
                    [
                        person,
                        person2,
                    ],
                    [
                        person3,
                        person4,
                    ],
                ],
            };

            const subject = shallow(<AddressBookTable {...props}/>);

            subject.setState({ page: props.maxPage });

            expect(subject.find(TableBody)).toMatchSnapshot();
        });
    });
});