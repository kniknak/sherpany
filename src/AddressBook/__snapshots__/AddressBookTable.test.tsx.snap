// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddressBook AddressBookTable renders 1`] = `
<Fragment>
  <WithStyles(ForwardRef(Table))>
    <WithStyles(ForwardRef(TableHead))>
      <WithStyles(ForwardRef(TableRow))>
        <WithStyles(ForwardRef(TableCell))>
          picture.thumbnail
        </WithStyles(ForwardRef(TableCell))>
        <WithStyles(ForwardRef(TableCell))>
          name.first
        </WithStyles(ForwardRef(TableCell))>
        <WithStyles(ForwardRef(TableCell))>
          name.last
        </WithStyles(ForwardRef(TableCell))>
        <WithStyles(ForwardRef(TableCell))>
          name.username
        </WithStyles(ForwardRef(TableCell))>
        <WithStyles(ForwardRef(TableCell))>
          email
        </WithStyles(ForwardRef(TableCell))>
        <WithStyles(ForwardRef(TableCell)) />
      </WithStyles(ForwardRef(TableRow))>
    </WithStyles(ForwardRef(TableHead))>
    <WithStyles(ForwardRef(TableBody)) />
  </WithStyles(ForwardRef(Table))>
  <Memo()>
    end of users catalog
  </Memo()>
  <AddressBookPersonModal
    isOpen={false}
    onClose={[Function]}
  />
</Fragment>
`;

exports[`AddressBook AddressBookTable renders end state 1`] = `
<Memo()>
  end of users catalog
</Memo()>
`;

exports[`AddressBook AddressBookTable renders loading state 1`] = `
<Memo()>
  loading...
</Memo()>
`;

exports[`AddressBook AddressBookTable renders table 1`] = `
<WithStyles(ForwardRef(TableBody))>
  <AddressBookPerson
    isShown={true}
    key="id.value"
    onOpen={[Function]}
    person={
      Object {
        "cell": "cell",
        "email": "email",
        "id": Object {
          "value": "id.value",
        },
        "location": Object {
          "city": "location.city",
          "postcode": "location.postcode",
          "state": "location.state",
          "street": Object {
            "name": "location.street.name",
            "number": 1,
          },
        },
        "login": Object {
          "username": "login.username",
        },
        "name": Object {
          "first": "name.first",
          "last": "name.last",
        },
        "phone": "phone",
        "picture": Object {
          "thumbnail": "picture.thumbnail",
        },
      }
    }
  />
  <AddressBookPerson
    isShown={true}
    key="id.value2"
    onOpen={[Function]}
    person={
      Object {
        "cell": "cell",
        "email": "email",
        "id": Object {
          "value": "id.value2",
        },
        "location": Object {
          "city": "location.city",
          "postcode": "location.postcode",
          "state": "location.state",
          "street": Object {
            "name": "location.street.name",
            "number": 1,
          },
        },
        "login": Object {
          "username": "login.username",
        },
        "name": Object {
          "first": "name.first",
          "last": "name.last",
        },
        "phone": "phone",
        "picture": Object {
          "thumbnail": "picture.thumbnail",
        },
      }
    }
  />
  <AddressBookPerson
    isShown={true}
    key="id.value2"
    onOpen={[Function]}
    person={
      Object {
        "cell": "cell",
        "email": "email",
        "id": Object {
          "value": "id.value2",
        },
        "location": Object {
          "city": "location.city",
          "postcode": "location.postcode",
          "state": "location.state",
          "street": Object {
            "name": "location.street.name",
            "number": 1,
          },
        },
        "login": Object {
          "username": "login.username",
        },
        "name": Object {
          "first": "name.first",
          "last": "name.last",
        },
        "phone": "phone",
        "picture": Object {
          "thumbnail": "picture.thumbnail",
        },
      }
    }
  />
  <AddressBookPerson
    isShown={true}
    key="id.value2"
    onOpen={[Function]}
    person={
      Object {
        "cell": "cell",
        "email": "email",
        "id": Object {
          "value": "id.value2",
        },
        "location": Object {
          "city": "location.city",
          "postcode": "location.postcode",
          "state": "location.state",
          "street": Object {
            "name": "location.street.name",
            "number": 1,
          },
        },
        "login": Object {
          "username": "login.username",
        },
        "name": Object {
          "first": "name.first",
          "last": "name.last",
        },
        "phone": "phone",
        "picture": Object {
          "thumbnail": "picture.thumbnail",
        },
      }
    }
  />
</WithStyles(ForwardRef(TableBody))>
`;
